internal/vault/vault.go

```diff
@@ -361,22 +376,56 @@ func (v *Vault) requestTokenWithAppRoleRef(client Client, appRole *v1.VaultAppRo
 }
 
 func (v *Vault) requestTokenWithKubernetesAuth(client Client, kubernetesAuth *v1.VaultKubernetesAuth) (string, error) {
-	secret, err := v.secretsLister.Secrets(v.namespace).Get(kubernetesAuth.SecretRef.Name)
-	if err != nil {
-		return "", err
-	}
+	var jwt string
+	switch {
+	case kubernetesAuth.SecretRef.Name != "":
+		secret, err := v.secretsLister.Secrets(v.namespace).Get(kubernetesAuth.SecretRef.Name)
+		if err != nil {
+			return "", err
+		}
 
-	key := kubernetesAuth.SecretRef.Key
-	if key == "" {
-		key = v1.DefaultVaultTokenAuthSecretKey
-	}
+		key := kubernetesAuth.SecretRef.Key
+		if key == "" {
+			key = v1.DefaultVaultTokenAuthSecretKey
+		}
 
-	keyBytes, ok := secret.Data[key]
-	if !ok {
-		return "", fmt.Errorf("no data for %q in secret '%s/%s'", key, v.namespace, kubernetesAuth.SecretRef.Name)
-	}
+		keyBytes, ok := secret.Data[key]
+		if !ok {
+			return "", fmt.Errorf("no data for %q in secret '%s/%s'", key, v.namespace, kubernetesAuth.SecretRef.Name)
+		}
+
+		jwt = string(keyBytes)
 
-	jwt := string(keyBytes)
+	case kubernetesAuth.ServiceAccountRef != nil:
+		aud := "vault://"
+		if v.issuer.GetNamespace() != "" {
+			aud += v.issuer.GetNamespace() + "/"
+		}
+		aud += v.issuer.GetName()
```

Comment:
The audience will look like this: `vault://namespace123/issuer123` or `vault://clusterissuer123`
I think this is a valid solution, alternatives that I would have also accepted/ preferred: a static `vault` audience.

Reply:
You are right, I reverted to the initial solution even though we had settled on using the hard-coded `vault`.

I added this bit of context to the PR description:

> **Alternative solution considered:** we considered using a hard-coded `vault` audience, but that did not allow to prevent the risk 1 mentioned in https://github.com/cert-manager/cert-manager/pull/5502#issuecomment-1337377810. The UX would have been better, and would correspond to what people expect from an audience (i.e., the audience should be commanded by Vault to cert-manager, not the other way around). But since this is an extra level of security (extra because Vault won't check the audience by default), it should not affect the UX much.

I hope that explains why I still went with the "generated audience".